powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/generator/IdentifierGenerator.java

```diff
@@ -220,17 +220,19 @@ public RecoveryInfo generateRecoveryCode(SecretKey secretKey, int pukCount, bool
         final Map<Integer, String> puks = new LinkedHashMap<>();
 
         for (int i = 1; i <= pukCount; i++) {
-            byte[] derivationIndexBytes;
             Long derivationIndex;
+            String derivedPuk;
             do {
-                // Generate random derivation index which must be unique
-                derivationIndexBytes = keyGenerator.generateRandomBytes(8);
+                // Generate random derivation index
+                byte[] derivationIndexBytes = keyGenerator.generateRandomBytes(8);
                 derivationIndex = ByteBuffer.wrap(derivationIndexBytes).getLong();
-            } while (pukDerivationIndexes.values().contains(derivationIndex));
+                // Generate recovery PUK
+                derivedPuk = generatePuk(recoveryPukBaseKey, derivationIndexBytes);
+                // Make sure that generated PUK is unique
+            } while (puks.values().contains(derivedPuk));
```

Comment:
There should be some cycle termination condition for when the universe is broken.

Reply:
Ok, I will add it.